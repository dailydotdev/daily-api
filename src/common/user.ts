import { DataSource } from 'typeorm';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';
import { ghostUser } from './index';
import { cancelSubscription } from './paddle';
import type { AuthContext, Context } from '../Context';
import { ForbiddenError } from 'apollo-server-errors';
import { logger } from '../logger';
import { CoresRole } from '../types';
import { remoteConfig } from '../remoteConfig';
import { UserTransaction } from '../entity/user/UserTransaction';
import { SubscriptionProvider, SubscriptionStatus } from './plus';
import { deleteUserResume } from './googleCloud';

export const deleteUser = async (
  con: DataSource,
  userId: string,
  messageId?: string,
) => {
  try {
    const { subscriptionFlags } = await con.getRepository(User).findOneOrFail({
      select: ['subscriptionFlags'],
      where: { id: userId },
    });

    // If the user has a subscription, we need to handle it before deletion
    if (subscriptionFlags?.subscriptionId) {
      if (
        subscriptionFlags?.provider === SubscriptionProvider.AppleStoreKit &&
        subscriptionFlags?.status === SubscriptionStatus.Active
      ) {
        throw new Error(
          'Apple subscriptions are not supported for user deletion',
        );
      }

      if (subscriptionFlags?.provider === SubscriptionProvider.Paddle) {
        const isGifted = !!subscriptionFlags.giftExpirationDate;
        // gifted subscription is a one-time payment hence not considered subscription in Paddle's terms
        if (!isGifted) {
          await cancelSubscription({
            subscriptionId: subscriptionFlags.subscriptionId,
          });
        }
        logger.info(
          {
            provider: SubscriptionProvider.Paddle,
            userId,
            isGifted,
            subscriptionId: subscriptionFlags.subscriptionId,
          },
          'Subscription cancelled user deletion',
        );
      }
    }

    // Delete user's resume if exists
    await deleteUserResume(userId);

    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: ghostUser.id,
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: ghostUser.id });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(UserTransaction).update(
        {
          senderId: userId,
        },
        {
          senderId: ghostUser.id,
        },
      );
      await entityManager.getRepository(UserTransaction).update(
        {
          receiverId: userId,
        },
        {
          receiverId: ghostUser.id,
        },
      );
      await entityManager.getRepository(User).delete(userId);
    });
    logger.info(
      {
        userId,
        messageId,
      },
      'deleted user',
    );
  } catch (err) {
    logger.error(
      {
        userId,
        messageId,
        err,
      },
      'failed to delete user',
    );
    throw err;
  }
};

export /**
 * Function creator that wraps a function with auth protection
 * In this case it checks ctx contains userId but it can be expanded
 * in the future to check more advanced things.
 * This is a DX layer to make sure auth required stuff like cores
 * is protected on the function level.
 *
 * @template Props
 * @template Result
 * @param {(props: Props) => Result} fn
 * @return {Result}
 */
const createAuthProtectedFn = <
  Props extends { ctx: Pick<AuthContext, 'userId'> },
  Result,
>(
  fn: (props: Props) => Result,
) => {
  return (props: Props): Result => {
    if (!props.ctx.userId) {
      throw new ForbiddenError('Auth is required');
    }

    return fn(props);
  };
};

export const getUserCoresRole = ({ region }: { region?: string }) => {
  if (!region) {
    return CoresRole.None;
  }

  const ruleForRegion = remoteConfig.vars.coresRoleRules?.find((rule) =>
    rule.regions.includes(region),
  );

  if (ruleForRegion) {
    return ruleForRegion.role;
  }

  return CoresRole.Creator;
};

export const checkUserCoresAccess = ({
  user,
  requiredRole,
}: {
  user: Pick<User, 'id' | 'coresRole'>;
  requiredRole: CoresRole;
}): boolean => {
  if (!user) {
    return false;
  }

  if (typeof user.coresRole !== 'number') {
    return false;
  }

  if (user.coresRole < requiredRole) {
    return false;
  }

  return true;
};

export const checkCoresAccess = async ({
  ctx,
  userId,
  requiredRole,
}: {
  ctx: Context;
  userId: string;
  requiredRole: CoresRole;
}): Promise<boolean> => {
  const user = await ctx.dataLoader.user.load({
    userId,
    select: ['id', 'coresRole'],
  });

  if (!user) {
    return false;
  }

  return checkUserCoresAccess({ user, requiredRole });
};
