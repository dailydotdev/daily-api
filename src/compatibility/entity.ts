import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
