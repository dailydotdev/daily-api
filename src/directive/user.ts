import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';

export const deleteUser = async (
  con: DataSource,
  logger: FastifyLoggerInstance,
  userId: string,
  messageId?: string,
) => {
  try {
    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: '404',
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: '404' });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(User).delete(userId);
    });
    if (logger) {
      logger.info(
        {
          userId,
          messageId,
        },
        'deleted user',
      );
    }
  } catch (err) {
    if (logger) {
      logger.error(
        {
          userId,
          messageId,
          err,
        },
        'failed to delete user',
      );
    }
    throw err;
  }
};
