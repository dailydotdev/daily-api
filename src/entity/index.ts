export * from './user';
export * from './AdvancedSettings';
export * from './Alerts';
export * from './Banner';
export * from './Bookmark';
export * from './BookmarkList';
export * from './DevCard';
export * from './Feed';
export * from './FeedAdvancedSettings';
export * from './FeedSource';
export * from './FeedTag';
export * from './Integration';
export * from './posts';
export * from './PostReport';
export * from './PostTag';
export * from './Settings';
export * from './Source';
export * from './SourceDisplay';
export * from './SourceFeed';
export * from './SourceMember';
export * from './SourceRequest';
export * from './TagSegment';
export * from './View';
export * from './Comment';
export * from './Keyword';
export * from './PostKeyword';
export * from './CommentMention';
export * from './ReputationEvent';
export * from './Submission';
export * from './notifications';
export * from './Feature';
export * from './Invite';
export * from './ContentImage';
export * from './MarketingCta';
export * from './SquadPublicRequest';
