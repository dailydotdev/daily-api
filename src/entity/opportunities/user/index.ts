export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
