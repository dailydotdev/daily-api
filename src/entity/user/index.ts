export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
