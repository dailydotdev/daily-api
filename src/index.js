import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import KoaPinoLogger from 'koa-pino-logger';
import Router from 'koa-router';
import etag from 'koa-etag';


import config from './config';
// import compress from './middlewares/compress';
import errorHandler from './middlewares/errorHandler';
import logger from './logger';

import health from './routes/health';
import sources from './routes/sources';
import posts from './routes/posts';
import publications from './routes/publications';
import tweet from './routes/tweet';
import settings from './routes/settings';
import feeds from './routes/feeds';
import notifications from './routes/notifications';
import tags from './routes/tags';

const app = new Koa();

app.keys = [config.cookies.secret];

app.proxy = true;

// TODO: disabled due to performance issues
// app.use(compress());

app.use(bodyParser());
app.use(KoaPinoLogger({ logger, useLevel: 'debug' }));
app.use(errorHandler());
app.use(etag());

// Machine-to-machine authentication
app.use((ctx, next) => {
  if (ctx.request.get('authorization') === `Service ${config.accessSecret}`
    && ctx.request.get('user-id') && ctx.request.get('logged-in')) {
    // eslint-disable-next-line
    ctx.state = {
      user: {
        userId: ctx.request.get('user-id'),
      },
      service: true,
    };
  } else {
    delete ctx.request.headers['user-id'];
    delete ctx.request.headers['logged-in'];
  }
  return next();
});

const router = new Router({
  prefix: '/v1',
});

router.use(feeds.routes(), feeds.allowedMethods());
router.use(sources.routes(), sources.allowedMethods());
router.use(posts.routes(), posts.allowedMethods());
router.use(publications.routes(), publications.allowedMethods());
router.use(tweet.routes(), tweet.allowedMethods());
router.use(settings.routes(), settings.allowedMethods());
router.use(notifications.routes(), notifications.allowedMethods());
router.use(tags.routes(), tags.allowedMethods());

app.use(router.routes(), router.allowedMethods());
app.use(health.routes(), health.allowedMethods());

export default app;
