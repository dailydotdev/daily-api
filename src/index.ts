import 'reflect-metadata';
import fastify, {
  FastifyRequest,
  FastifyInstance,
  FastifyError,
} from 'fastify';
import fastifyRawBody from 'fastify-raw-body';
import helmet from '@fastify/helmet';
import cors from '@fastify/cors';
import mercurius, { MercuriusError } from 'mercurius';
import MercuriusGQLUpload from 'mercurius-upload';
import MercuriusCache from 'mercurius-cache';
import { NoSchemaIntrospectionCustomRule } from 'graphql';
// import fastifyWebsocket from '@fastify/websocket';

import './config';

import tracking from './tracking';
import auth from './auth';
import compatibility from './compatibility';
import routes from './routes';
import { Context } from './Context';
import { schema } from './graphql';
import createOrGetConnection from './db';
import { stringifyHealthCheck } from './common';
import { GraphQLError } from 'graphql';
import cookie, { FastifyCookieOptions } from '@fastify/cookie';
import { getSubscriptionSettings } from './subscription';
import { ioRedisPool } from './redis';
import { loadFeatures } from './growthbook';
import { runInRootSpan } from './telemetry/opentelemetry';
import { loggerConfig } from './logger';

type Mutable<Type> = {
  -readonly [Key in keyof Type]: Type[Key];
};

const userExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined => (ctx.userId ? `user:${ctx.userId}` : undefined);

const trackingExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined =>
  ctx.trackingId ? `tracking:${ctx.trackingId}` : undefined;

// readiness probe is set failureThreshold: 2, periodSeconds: 2 (4s) + small delay
const GRACEFUL_DELAY = 2 * 2 * 1000 + 5000;

export default async function app(
  contextFn?: (request: FastifyRequest) => Context,
): Promise<FastifyInstance> {
  let isTerminating = false;
  const isProd = process.env.NODE_ENV === 'production';
  const connection = await runInRootSpan(
    'createOrGetConnection',
    createOrGetConnection,
  );

  const app = fastify({
    logger: loggerConfig,
    disableRequestLogging: true,
    trustProxy: true,
  });

  app.log.info('loading features');
  await loadFeatures(app.log);

  const gracefulShutdown = () => {
    app.log.info('starting termination');
    isTerminating = true;
    setTimeout(async () => {
      await app.close();
      await connection.destroy();
      await ioRedisPool.end();
      process.exit();
    }, GRACEFUL_DELAY);
  };
  process.on('SIGINT', gracefulShutdown);
  process.on('SIGTERM', gracefulShutdown);

  app.register(helmet);
  app.register(cors, {
    origin: isProd ? /daily\.dev$/ : true,
    credentials: true,
  });
  app.register(cookie, {
    secret: process.env.COOKIES_KEY,
  }) as FastifyCookieOptions;
  app.register(auth, { secret: process.env.ACCESS_SECRET });
  app.register(tracking);
  app.register(fastifyRawBody, {
    field: 'rawBody',
    global: false,
    encoding: 'utf8',
    runFirst: true,
  });

  app.setErrorHandler((err, req, res) => {
    req.log.error({ err }, err.message);
    res.code(500).send({ statusCode: 500, error: 'Internal Server Error' });
  });

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    if (isTerminating) {
      res.status(500).send(stringifyHealthCheck({ status: 'terminating' }));
    } else {
      res.send(stringifyHealthCheck({ status: 'ok' }));
    }
  });

  app.get('/liveness', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  // app.register(fastifyWebsocket, {
  //   options: {
  //     maxPayload: 1048576,
  //     verifyClient: (info, next) => next(true),
  //   },
  // });

  app.register(MercuriusGQLUpload, {
    maxFileSize: 1024 * 1024 * 20,
    maxFiles: 1,
  });

  app.register(mercurius, {
    schema,
    context:
      contextFn ?? ((request): Context => new Context(request, connection)),
    queryDepth: 10,
    subscription: getSubscriptionSettings(connection),
    // Disable GraphQL introspection in production
    graphiql: !isProd,
    validationRules: isProd && [NoSchemaIntrospectionCustomRule],
    errorFormatter(execution, ctx) {
      if (execution.errors?.length > 0) {
        const flatErrors = execution.errors.flatMap<GraphQLError>((error) => {
          if (error.originalError.name === 'FastifyError') {
            return (
              (error.originalError as MercuriusError<GraphQLError>).errors ||
              error
            );
          }
          return error;
        });
        return {
          statusCode: 200,
          response: {
            data: execution.data,
            errors: flatErrors.map((error): GraphQLError => {
              const newError = error as Mutable<GraphQLError>;
              if (!error.originalError) {
                newError.extensions = {
                  code: 'GRAPHQL_VALIDATION_FAILED',
                };
              } else if (error.originalError?.name === 'EntityNotFoundError') {
                newError.message = 'Entity not found';
                newError.extensions = {
                  code: 'NOT_FOUND',
                };
              } else if (
                (error.originalError as FastifyError)?.code ===
                'MER_ERR_GQL_PERSISTED_QUERY_NOT_FOUND'
              ) {
                app.log.debug(
                  { body: ctx?.reply?.request?.body },
                  'unknown query',
                );
              } else if (!error.extensions?.code) {
                app.log.warn(
                  {
                    err: error.originalError,
                    body: ctx?.reply?.request?.body,
                  },
                  'unexpected graphql error',
                );
                newError.message = 'Unexpected error';
                newError.extensions = {
                  code: 'UNEXPECTED',
                };
              }
              if (isProd) {
                newError.originalError = undefined;
              }
              return newError;
            }),
          },
        };
      }
      return {
        statusCode: 200,
        response: execution,
      };
    },
  });

  if (isProd) {
    app.register(MercuriusCache, {
      ttl: 10,
      policy: {
        Query: {
          searchBookmarksSuggestions: {
            extendKey: userExtendKey,
          },
          searchBookmarks: {
            extendKey: userExtendKey,
          },
          // anonymousFeed: {
          //   extendKey: userExtendKey,
          // },
          // feed: {
          //   extendKey: userExtendKey,
          // },
          sourceFeed: {
            extendKey: userExtendKey,
          },
          tagFeed: {
            extendKey: userExtendKey,
          },
          keywordFeed: {
            extendKey: userExtendKey,
          },
          searchPostSuggestions: {
            extendKey: userExtendKey,
          },
          searchPosts: {
            extendKey: userExtendKey,
          },
          authorFeed: {
            extendKey: userExtendKey,
          },
          mostUpvotedFeed: {
            extendKey: userExtendKey,
          },
          mostDiscussedFeed: {
            extendKey: userExtendKey,
          },
          randomTrendingPosts: {
            extendKey: userExtendKey,
          },
          randomSimilarPostsByTags: {
            extendKey: userExtendKey,
          },
          randomDiscussedPosts: {
            extendKey: userExtendKey,
          },
          tagsCategories: true,
          advancedSettings: {
            extendKey: trackingExtendKey,
          },
          banner: true,
          postByUrl: {
            extendKey: userExtendKey,
          },
          post: {
            extendKey: userExtendKey,
          },
          postUpvotes: true,
          searchTags: true,
          userReadingRankHistory: true,
          userReadHistory: true,
        },
      },
    });
  }

  app.register(compatibility, { prefix: '/v1' });
  app.register(routes, { prefix: '/' });

  return app;
}
