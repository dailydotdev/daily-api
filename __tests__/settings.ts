import { FastifyInstance } from 'fastify';
import request from 'supertest';
import { v4 as uuidv4 } from 'uuid';
import {
  authorizeRequest,
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
  testMutationErrorCode,
  testQueryError,
  testQueryErrorCode,
} from './helpers';
import { CampaignCtaPlacement, Settings, User } from '../src/entity';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../src/db';
import { usersFixture } from './fixture/user';

let app: FastifyInstance;
let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser),
  );
  client = state.client;
  app = state.app;
});

beforeEach(async () => {
  loggedUser = null;

  await saveFixtures(con, User, usersFixture);
});

afterAll(() => disposeGraphQLTesting(state));

const compatibilityProps = { enableCardAnimations: true, appInsaneMode: true };

describe('query userSettings', () => {
  const QUERY = `{
  userSettings {
    userId
    theme
    enableCardAnimations
    showTopSites
    insaneMode
    appInsaneMode
    spaciness
    showOnlyUnreadPosts
    openNewTab
    sidebarExpanded
    companionExpanded
    sortingEnabled
    customLinks
    optOutWeeklyGoal
    optOutReadingStreak
    optOutCompanion
    autoDismissNotifications
    campaignCtaPlacement
  }
}`;

  it('should not authorize when not logged-in', () =>
    testQueryErrorCode(client, { query: QUERY }, 'UNAUTHENTICATED'));

  it('should return user settings', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    const settings = repo.create({
      userId: '1',
      theme: 'bright',
      insaneMode: true,
      campaignCtaPlacement: CampaignCtaPlacement.Header,
    });
    const data = await repo.save(settings);
    const expected = new Object({ ...data, ...compatibilityProps });
    delete expected['updatedAt'];
    delete expected['bookmarkSlug'];

    const res = await client.query(QUERY);
    expect(res.data.userSettings).toEqual(expected);
  });

  it('should create default settings if not exist', async () => {
    loggedUser = '1';
    const res = await client.query(QUERY);
    expect(res.data.userSettings).toMatchSnapshot();
  });
});

describe('query bookmarksSharing', () => {
  const QUERY = `{
  bookmarksSharing {
    enabled
    slug
    rssUrl
  }
}`;

  it('should not authorize when not logged-in', () =>
    testQueryErrorCode(client, { query: QUERY }, 'UNAUTHENTICATED'));

  it('should return disabled settings', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    const settings = repo.create({ userId: '1' });
    const data = await repo.save(settings);
    const expected = new Object({ ...data, ...compatibilityProps });
    delete expected['updatedAt'];

    const res = await client.query(QUERY);
    expect(res.data.bookmarksSharing.enabled).toEqual(false);
    expect(res.data.bookmarksSharing.slug).toBeFalsy();
    expect(res.data.bookmarksSharing.rssUrl).toBeFalsy();
  });

  it('should return enabled settings', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    const settings = repo.create({ userId: '1', bookmarkSlug: uuidv4() });
    const data = await repo.save(settings);
    const expected = new Object({ ...data, ...compatibilityProps });
    delete expected['updatedAt'];

    const res = await client.query(QUERY);
    expect(res.data.bookmarksSharing.enabled).toEqual(true);
    expect(res.data.bookmarksSharing.slug).toEqual(settings.bookmarkSlug);
    expect(res.data.bookmarksSharing.rssUrl).toEqual(
      `http://localhost:4000/rss/b/${settings.bookmarkSlug}`,
    );
  });
});

describe('mutation updateUserSettings', () => {
  const MUTATION = `
  mutation UpdateUserSettings($data: UpdateSettingsInput!) {
  updateUserSettings(data: $data) {
    userId
    theme
    enableCardAnimations
    showTopSites
    insaneMode
    appInsaneMode
    spaciness
    showOnlyUnreadPosts
    openNewTab
    sidebarExpanded
    companionExpanded
    sortingEnabled
    customLinks
    optOutWeeklyGoal
    optOutReadingStreak
    optOutCompanion
    campaignCtaPlacement
  }
}`;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
        variables: { data: { theme: 'bright', insaneMode: true } },
      },
      'UNAUTHENTICATED',
    ));

  it('should create user settings when does not exist', async () => {
    loggedUser = '1';
    const res = await client.mutate(MUTATION, {
      variables: { data: { theme: 'bright', insaneMode: true } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should not allow invalid user links', async () => {
    loggedUser = '1';

    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
        variables: { data: { customLinks: ['http://'] } },
      },
      'GRAPHQL_VALIDATION_FAILED',
    );
  });

  it('should not allow invalid user links even when there are valid ones', async () => {
    loggedUser = '1';

    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
        variables: {
          data: { customLinks: ['https://app.daily.dev', 'http://'] },
        },
      },
      'GRAPHQL_VALIDATION_FAILED',
    );
  });

  it('should update user links if valid', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    await repo.save(
      repo.create({
        userId: '1',
      }),
    );

    const res = await client.mutate(MUTATION, {
      variables: { data: { customLinks: ['http://abc.com'] } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should update user settings', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    await repo.save(
      repo.create({
        userId: '1',
        theme: 'bright',
        insaneMode: true,
        customLinks: ['http://abc.com'],
      }),
    );

    const res = await client.mutate(MUTATION, {
      variables: { data: { insaneMode: false } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should update opt out companion settings', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    await repo.save(
      repo.create({
        userId: '1',
        optOutCompanion: false,
      }),
    );

    const res = await client.mutate(MUTATION, {
      variables: { data: { optOutCompanion: true } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should update campaignCtaPlacement', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    await repo.insert({
      userId: '1',
      optOutCompanion: false,
      campaignCtaPlacement: CampaignCtaPlacement.Header,
    });

    const res = await client.mutate(MUTATION, {
      variables: {
        data: { campaignCtaPlacement: CampaignCtaPlacement.ProfileMenu },
      },
    });

    expect(res.data.updateUserSettings.campaignCtaPlacement).toBe(
      CampaignCtaPlacement.ProfileMenu,
    );
  });

  it('should return a validation error if passed an invalid value for campaignCtaPlacement', async () => {
    loggedUser = '1';

    const repo = con.getRepository(Settings);
    await repo.insert({
      userId: '1',
      optOutCompanion: false,
      campaignCtaPlacement: CampaignCtaPlacement.Header,
    });

    await testQueryError(
      client,
      {
        query: MUTATION,
        variables: {
          data: { campaignCtaPlacement: 'foo' as CampaignCtaPlacement },
        },
      },
      (errors) => {
        expect(errors[0].extensions.code).toEqual('GRAPHQL_VALIDATION_FAILED');
        expect(errors[0].message).toEqual(
          `Invalid value for 'campaignCtaPlacement'`,
        );
      },
    );
  });
});

describe('mutation setBookmarksSharing', () => {
  const MUTATION = `
  mutation SetBookmarksSharing($enabled: Boolean!) {
  setBookmarksSharing(enabled: $enabled) {
    enabled
    slug
    rssUrl
  }
}`;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      {
        mutation: MUTATION,
        variables: { enabled: true },
      },
      'UNAUTHENTICATED',
    ));

  it('should enable bookmarks sharing', async () => {
    loggedUser = '1';
    const res = await client.mutate(MUTATION, {
      variables: { enabled: true },
    });
    expect(res.data.setBookmarksSharing.enabled).toEqual(true);
    expect(res.data.setBookmarksSharing.slug).toBeTruthy();
    expect(res.data.setBookmarksSharing.rssUrl).toEqual(
      `http://localhost:4000/rss/b/${res.data.setBookmarksSharing.slug}`,
    );
  });

  it('should disable bookmarks sharing', async () => {
    loggedUser = '1';
    const repo = con.getRepository(Settings);
    const settings = repo.create({ userId: '1', bookmarkSlug: uuidv4() });
    await repo.save(settings);
    const res = await client.mutate(MUTATION, {
      variables: { enabled: false },
    });
    expect(res.data.setBookmarksSharing.enabled).toEqual(false);
    expect(res.data.setBookmarksSharing.slug).toBeFalsy();
    expect(res.data.setBookmarksSharing.rssUrl).toBeFalsy();
  });

  it('should not do anything when bookmarks sharing is enabled', async () => {
    loggedUser = '1';
    const repo = con.getRepository(Settings);
    const settings = repo.create({ userId: '1', bookmarkSlug: uuidv4() });
    const data = await repo.save(settings);
    const res = await client.mutate(MUTATION, {
      variables: { enabled: true },
    });
    expect(res.data.setBookmarksSharing.enabled).toEqual(true);
    expect(res.data.setBookmarksSharing.slug).toEqual(data.bookmarkSlug);
  });
});

describe('dedicated api routes', () => {
  describe('GET /settings', () => {
    it('should return user settings', async () => {
      const repo = con.getRepository(Settings);
      const settings = repo.create({
        userId: '1',
        theme: 'bright',
        insaneMode: true,
      });
      const data = await repo.save(settings);
      const expected = new Object(data);
      delete expected['updatedAt'];
      delete expected['userId'];
      delete expected['bookmarkSlug'];

      loggedUser = '1';
      const res = await authorizeRequest(
        request(app.server).get('/settings'),
      ).expect(200);
      expect(res.body).toEqual(expected);
    });
  });
});
