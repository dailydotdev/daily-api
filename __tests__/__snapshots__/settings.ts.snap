// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation updateUserSettings should create user settings when does not exist 1`] = `
Object {
  "updateUserSettings": Object {
    "appInsaneMode": true,
    "campaignCtaPlacement": "header",
    "companionExpanded": null,
    "customLinks": null,
    "enableCardAnimations": true,
    "insaneMode": true,
    "openNewTab": true,
    "optOutCompanion": false,
    "optOutReadingStreak": false,
    "optOutWeeklyGoal": false,
    "showOnlyUnreadPosts": false,
    "showTopSites": true,
    "sidebarExpanded": true,
    "sortingEnabled": false,
    "spaciness": "eco",
    "theme": "bright",
    "userId": "1",
  },
}
`;

exports[`mutation updateUserSettings should update opt out companion settings 1`] = `
Object {
  "updateUserSettings": Object {
    "appInsaneMode": true,
    "campaignCtaPlacement": "header",
    "companionExpanded": null,
    "customLinks": null,
    "enableCardAnimations": true,
    "insaneMode": false,
    "openNewTab": true,
    "optOutCompanion": true,
    "optOutReadingStreak": false,
    "optOutWeeklyGoal": false,
    "showOnlyUnreadPosts": false,
    "showTopSites": true,
    "sidebarExpanded": true,
    "sortingEnabled": false,
    "spaciness": "eco",
    "theme": "darcula",
    "userId": "1",
  },
}
`;

exports[`mutation updateUserSettings should update user links if valid 1`] = `
Object {
  "updateUserSettings": Object {
    "appInsaneMode": true,
    "campaignCtaPlacement": "header",
    "companionExpanded": null,
    "customLinks": Array [
      "http://abc.com",
    ],
    "enableCardAnimations": true,
    "insaneMode": false,
    "openNewTab": true,
    "optOutCompanion": false,
    "optOutReadingStreak": false,
    "optOutWeeklyGoal": false,
    "showOnlyUnreadPosts": false,
    "showTopSites": true,
    "sidebarExpanded": true,
    "sortingEnabled": false,
    "spaciness": "eco",
    "theme": "darcula",
    "userId": "1",
  },
}
`;

exports[`mutation updateUserSettings should update user settings 1`] = `
Object {
  "updateUserSettings": Object {
    "appInsaneMode": true,
    "campaignCtaPlacement": "header",
    "companionExpanded": null,
    "customLinks": Array [
      "http://abc.com",
    ],
    "enableCardAnimations": true,
    "insaneMode": false,
    "openNewTab": true,
    "optOutCompanion": false,
    "optOutReadingStreak": false,
    "optOutWeeklyGoal": false,
    "showOnlyUnreadPosts": false,
    "showTopSites": true,
    "sidebarExpanded": true,
    "sortingEnabled": false,
    "spaciness": "eco",
    "theme": "bright",
    "userId": "1",
  },
}
`;

exports[`query userSettings should create default settings if not exist 1`] = `
Object {
  "appInsaneMode": true,
  "autoDismissNotifications": true,
  "campaignCtaPlacement": "header",
  "companionExpanded": false,
  "customLinks": null,
  "enableCardAnimations": true,
  "insaneMode": false,
  "openNewTab": true,
  "optOutCompanion": false,
  "optOutReadingStreak": false,
  "optOutWeeklyGoal": false,
  "showOnlyUnreadPosts": false,
  "showTopSites": true,
  "sidebarExpanded": true,
  "sortingEnabled": false,
  "spaciness": "eco",
  "theme": "darcula",
  "userId": "1",
}
`;
